### ISSUE TITLE
StaggeredGridLayoutManager

### ISSUE THREAD

**Issue by AcgnCodeMonkey** [Fondesa/RecyclerViewDivider#36]
![QQ截图20191023160222](https://user-images.githubusercontent.com/19161307/67371200-938b3800-f5ae-11e9-8151-cca6515ccf05.png)
Horizontal and Vertical space size is not euqal.
 RecyclerViewDivider.with(rv.getContext())
                .insetManager(new DefaultInsetManager())
                .hideLastDivider()
                .size(size)
                .color(ResUtils.getColor(color))
                .build().addTo(rv);

**Comment by AcgnCodeMonkey** [comment]
![QQ截图20191023160611](https://user-images.githubusercontent.com/19161307/67371497-0e545300-f5af-11e9-9723-f8ed93dfdce7.png)
left and right space also not equal.

**Comment by AcgnCodeMonkey** [comment]
but if i use  GridLayoutManager,those questions is ok.

**Comment by Fondesa** [comment]
Interesting. Guess I found the problem calculating the group count of a `StaggeredGridLayoutManager`. Is your project open source so I can directly test my changes there? Otherwise I'll try to reproduce your layout

**Comment by AcgnCodeMonkey** [comment]
> Interesting. Guess I found the problem calculating the group count of a `StaggeredGridLayoutManager`. Is your project open source so I can directly test my changes there? Otherwise I'll try to reproduce your layout

sorry,my project is enterprise project，i think you can write a simple demo to test this bug

**Comment by Fondesa** [comment]
Ok no problem, I'll keep you up-to-date

**Comment by AcgnCodeMonkey** [comment]
> Ok no problem, I'll keep you up-to-date

https://blog.csdn.net/qq_33373648/article/details/77584995
you can consult this blog.

**Comment by Fondesa** [comment]
This PR should fix it: https://github.com/Fondesa/RecyclerViewDivider/pull/37

**Comment by Fondesa** [comment]
Update to **3.0.2** and feel free to reopen the issue if it's not solved for you yet.

**Comment by AcgnCodeMonkey** [comment]
![QQ截图20191024160247](https://user-images.githubusercontent.com/19161307/67465467-d238f500-f677-11e9-8490-71b7cdcd7f14.png)
![QQ截图20191024160319](https://user-images.githubusercontent.com/19161307/67465468-d2d18b80-f677-11e9-80da-896a436416d9.png)
first page is normal,but second page is wrong

**Comment by AcgnCodeMonkey** [comment]
@Fondesa 
this is a demo.
[Demo.zip](https://github.com/Fondesa/RecyclerViewDivider/files/3766350/Demo.zip)

**Comment by AcgnCodeMonkey** [comment]
![QQ截图20191024160319](https://user-images.githubusercontent.com/19161307/67465673-2cd25100-f678-11e9-9854-8979b00696bc.png)

**Comment by Fondesa** [comment]
Ok thanks, I'll look into it

**Comment by Fondesa** [comment]
@AcgnCodeMonkey I found the problem.
It's related to the internal usage of `parent.getChildAdapterPosition(child)` which, with `StaggeredGridLayoutManager`, doesn't match the position in which the item is rendered.

This will be harder than I thought to fix it since the "entire" library is based on the item's position. 
Probably a solution would be to limit the position-dependent features of this library only for `StaggeredGridLayoutManager`.

**Comment by AcgnCodeMonkey** [comment]
because my kotlin is careless，i try do it. 

//        val lineAccumulatedSpan = lm.getAccumulatedSpanInLine(spanSize, itemPosition, groupIndex)
        var lineAccumulatedSpan = 0
        if (parent.layoutManager is StaggeredGridLayoutManager) {
            val params = view.layoutParams as StaggeredGridLayoutManager.LayoutParams
            lineAccumulatedSpan = params.spanIndex % spanCount + 1
        } else {
            lineAccumulatedSpan = lm.getAccumulatedSpanInLine(spanSize, itemPosition, groupIndex)
        }
and test result as follows :
![QQ截图20191028153352](https://user-images.githubusercontent.com/19161307/67659919-85fff480-f998-11e9-8b72-383b6b0b90e5.png)
![QQ截图20191028153447](https://user-images.githubusercontent.com/19161307/67659935-8f895c80-f998-11e9-9ce2-9ef7b3b7ea8a.png)
i check most pages ,all page is right.
i think you can draw lessons from that.

**Comment by AcgnCodeMonkey** [comment]
@Fondesa

**Comment by Fondesa** [comment]
@AcgnCodeMonkey Yes but this is not the problem. Your code works only if there's a basic configuration of the divider (e.g. all the dividers have the same size).
If the dividers have different configurations (e.g. they change the size based on `groupIndex`), the layout manager will render a wrong grid since the `groupIndex` isn't calculated right.

**Comment by Fondesa** [comment]
@AcgnCodeMonkey I've probably found a good solution for it. I'm working on it.

**Comment by Fondesa** [comment]
Here we go: https://github.com/Fondesa/RecyclerViewDivider/pull/39

The problem was related to the item position.
Since `StaggeredGridLayoutManager` can change the position of the items without change their adapter positions, I had to add a partial support to it cause all the managers were currently based on `groupIndex` and `groupCount`. Now I added a list of managers not dependent to those two variables.
I checked your sample and it works.
